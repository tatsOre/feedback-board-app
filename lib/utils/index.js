export const toCapitalize = (text) => {
  if (!text || typeof text !== 'string') return ''
  let label = text
  if (text.includes('-')) label = text.replace('-', ' ')
  return label
    .split(' ')
    .map((w) => w.charAt(0).toUpperCase() + w.slice(1))
    .join(' ')
}

export const getCommentsLength = (comments) => {
  if (!comments || !comments.length) return 0
  return comments.reduce((acc, comment) => {
    return acc + (comment.replies?.length || 0)
  }, comments.length || 0)
}

export const getRandomNumber = (min, max) => {
  let difference = max - min

  let rand = Math.random()

  rand = Math.floor(rand * difference)

  return rand + min
}

export const reduceFeedbacksData = (data) => {
  const accSchema = {
    planned: [],
    'in-progress': [],
    live: [],
    suggestion: [],
  }

  const result = data.reduce((acc, req) => {
    let status = req.status
    let doc = {
      id: req._id.toString(),
      title: req.title,
      slug: req.slug,
      description: req.description,
      status: req.status,
      category: req.category,
      comments: getCommentsLength(req.comments),
      upvotes: req.upvotes,
    }
    acc[status] = acc[status].length ? [...acc[status], doc] : [doc]
    return acc
  }, accSchema)

  return result
}
