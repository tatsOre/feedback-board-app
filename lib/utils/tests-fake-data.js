import MOCK_FEEDBACKS_DATA from '../../__mocks__/mock-feedbacks.json'
import MOCK_USER_DATA from '../../__mocks__/mock-user-info.json'
import MOCK_USER_FEEDBACK_DATA from '../../__mocks__/mock-user-feedback.json'
import MOCK_DETAILED_FEEDBACK_DATA from '../../__mocks__/mock-detailed-feedbacks.json'


export const MOCK_USER = MOCK_USER_DATA.data

export const MOCK_USER_FEEDBACK = MOCK_USER_FEEDBACK_DATA.data

export const MOCK_FEEDBACKS = MOCK_FEEDBACKS_DATA.data

export const MOCK_PLANNED = MOCK_FEEDBACKS.planned

export const MOCK_IN_PROGRESS = MOCK_FEEDBACKS['in-progress']

export const MOCK_LIVE = MOCK_FEEDBACKS.live

export const MOCK_SUGGESTION = MOCK_FEEDBACKS.suggestion

export const MOCK_DETAILED_FEEDBACK = MOCK_DETAILED_FEEDBACK_DATA.data


// import '@testing-library/jest-dom'