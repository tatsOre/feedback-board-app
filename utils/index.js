export const toCapitalize = (text) => {
  if (!text || typeof text !== 'string') return ''
  let label = text
  if (text.includes('-')) label = text.replace('-', ' ')
  return label
    .split(' ')
    .map((w) => w.charAt(0).toUpperCase() + w.slice(1))
    .join(' ')
}

export const getCommentsLength = (comments) => {
  if (!comments.length) return 0
  return comments.reduce((acc, comment) => {
    return acc + (comment.replies?.length || 0)
  }, comments.length || 0)
}

export const getRandomNumber = (min, max) => {
  let difference = max - min

  let rand = Math.random()

  rand = Math.floor(rand * difference)

  return rand + min
}
