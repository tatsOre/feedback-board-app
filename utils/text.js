export const toCapitalize = (text) => {
  if (!text || typeof text !== "string") return "";

  let label = text;
  if (text.includes("-")) label = text.replace("-", " ");

  return label
    .split(" ")
    .map((w) => w.charAt(0).toUpperCase() + w.slice(1))
    .join(" ");
};
