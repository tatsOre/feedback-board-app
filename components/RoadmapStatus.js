import Link from 'next/link'
import { Fragment } from 'react'

import { toCapitalize } from 'lib/utils'

export default function RoadmapStatus({ data }) {
  return (
    <article className="roadmap-status-panel bg-white rounded-10 w-full md:w-1/2 lg:w-full p-5 md:ml-2.5 lg:ml-0">
      <h2 className="text-indigo-800 text-lg inline-block mb-5">Roadmap</h2>
      <Link href="/roadmap" passHref>
        <a className="float-right text-blue-900 text-sm font-semibold underline mt-1">
          View
        </a>
      </Link>
      <dl className="flex flex-wrap items-baseline text-indigo-500">
        {Object.keys(data).map(
          (status) =>
            status !== 'suggestion' && (
              <Fragment key={`roadmap-status-${status}`}>
                <dt className={`title ${status} basis-3/4`}>
                  {toCapitalize(status)}
                </dt>
                <dd className="font-bold basis-1/4 text-right">
                  {data[status].length}
                </dd>
              </Fragment>
            )
        )}
      </dl>
    </article>
  )
}
