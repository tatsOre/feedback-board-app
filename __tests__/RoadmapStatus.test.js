import { render, screen } from '@testing-library/react'
import RoadmapStatus from '../components/RoadmapStatus'
import { MOCK_FEEDBACKS as data } from '../lib/utils/tests-fake-data'

describe('<RoadmapStatus />', () => {
  it('renders with correct data and matches snapshot', () => {
    const { container } = render(<RoadmapStatus data={data} />)

    expect(container).toMatchSnapshot()
  })

  it('renders link to roadmap page and correct tabular values', () => {
    render(<RoadmapStatus data={data} />)

    expect(screen.getByRole('link')).toHaveAttribute('href', '/roadmap')

    const planned = screen.getByText(/planned/i)

    expect(planned).toBeInTheDocument()

    expect(planned.nextElementSibling).toHaveTextContent(data.planned.length)

    const progress = screen.getByText(/in progress/i)

    expect(progress).toBeInTheDocument()

    expect(progress.nextElementSibling).toHaveTextContent(
      data['in-progress'].length
    )

    const live = screen.getByText(/live/i)

    expect(live).toBeInTheDocument()

    expect(live.nextElementSibling).toHaveTextContent(data.live.length)
  })
})
