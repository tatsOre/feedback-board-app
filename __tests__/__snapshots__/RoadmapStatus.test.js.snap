// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RoadmapStatus /> renders with correct data and matches snapshot 1`] = `
<div>
  <article
    class="roadmap-status-panel bg-white rounded-10 w-full md:w-1/2 lg:w-full p-5 md:ml-2.5 lg:ml-0"
  >
    <h2
      class="text-indigo-800 text-lg inline-block mb-5"
    >
      Roadmap
    </h2>
    <a
      class="float-right text-blue-900 text-sm font-semibold underline mt-1"
      href="/roadmap"
    >
      View
    </a>
    <dl
      class="flex flex-wrap items-baseline text-indigo-500"
    >
      <dt
        class="title planned basis-3/4"
      >
        Planned
      </dt>
      <dd
        class="font-bold basis-1/4 text-right"
      >
        2
      </dd>
      <dt
        class="title in-progress basis-3/4"
      >
        In Progress
      </dt>
      <dd
        class="font-bold basis-1/4 text-right"
      >
        3
      </dd>
      <dt
        class="title live basis-3/4"
      >
        Live
      </dt>
      <dd
        class="font-bold basis-1/4 text-right"
      >
        1
      </dd>
    </dl>
  </article>
</div>
`;
